urllib3/util/ssl_.py

```diff
@@ -323,19 +330,42 @@ def ssl_wrap_socket(sock, keyfile=None, certfile=None, cert_reqs=None,
         # try to load OS default certs; works well on Windows (require Python3.4+)
         context.load_default_certs()
 
+    # Due to a bug in Python's handling of SNI extensions (sends IP as SNI hostname when connecting via IP address)
+    # we shouldn't send things that parse as IP addresses to wrap_socket as a hostname.
+    # If ipaddress is available, use it to filter those out.
+    server_hostname_is_ip = False
+
+    if HAS_IPADDRESS:
+        try:
+            # ipaddress doesn't accept byte strings
+            if isinstance(server_hostname, u''.__class__):
+                server_hostname_uni = server_hostname
+            else:
+                server_hostname_uni = server_hostname.decode('utf-8')
+
+            ipaddress.ip_address(server_hostname_uni)
+            server_hostname_is_ip = True
+        except ipaddress.AddressValueError:
+            pass
+
     if certfile:
         context.load_cert_chain(certfile, keyfile)
+
+    wrap_socket_kwargs = {}
+
     if HAS_SNI:  # Platform-specific: OpenSSL with enabled SNI
-        return context.wrap_socket(sock, server_hostname=server_hostname)
-
-    warnings.warn(
-        'An HTTPS request has been made, but the SNI (Server Name '
-        'Indication) extension to TLS is not available on this platform. '
-        'This may cause the server to present an incorrect TLS '
-        'certificate, which can cause validation failures. You can upgrade to '
-        'a newer version of Python to solve this. For more information, see '
-        'https://urllib3.readthedocs.io/en/latest/advanced-usage.html'
-        '#ssl-warnings',
-        SNIMissingWarning
-    )
-    return context.wrap_socket(sock)
+        if not server_hostname_is_ip:
+            wrap_socket_kwargs['server_hostname'] = server_hostname
```

Comment:
You need to do hostname verification yourself if you do not supply server_hostname. Otherwise you do not check if the IP address is valid for the server's cert.

Reply:
Yeah; we would need a solution that 1. prevents the IP address from being sent pre-emptively as part of SNI but 2. does check the eventual received cert.